synapseclient/table.py

```diff
@@ -780,6 +781,125 @@ def __init__(self, name=None, columns=None, parent=None, definingSQL=None, prope
+    :py:classmethod:`synapseclient.client.create_snapshot_version`.
+
+    Example::
+        syn = synapseclient.login()
+        syn.create_snapshot_version(
+            dataset.id,
+            label="v1.0",
+            comment="This is version 1")
+    """
+    _synapse_entity_type: str = "org.sagebionetworks.repo.model.table.Dataset"
+    _property_keys: List[str] = SchemaBase._property_keys + ['dataset_items']
+
+    def __init__(self, name=None, columns=None, parent=None, properties=None,
+                 annotations=None, local_state=None, dataset_items=None, **kwargs):
+        self.properties.setdefault('dataset_items', [])
+        super(Dataset, self).__init__(
+            name=name, columns=columns, properties=properties,
+            annotations=annotations, local_state=local_state, parent=parent,
+            **kwargs
+        )
+        if dataset_items:
+            self.add_items(dataset_items)
+
+    def __len__(self):
+        return len(self.properties.dataset_items)
+
+    def add_item(self, dataset_item: Dict[str, str]):
+        """
+        :param dataset_item: a single dataset item
+        """
+        if isinstance(dataset_item, dict):
+            required_keys = {'entityId', 'versionNumber'}
+            if required_keys == dataset_item.keys():
+                self.properties.dataset_items.append(dataset_item)
+            elif required_keys - dataset_item.keys():
+                raise LookupError("DatasetItem missing a required property: %s" %
+                                  str(required_keys - dataset_item.keys()))
+
+            # FIXME? Having other key-value pair in the dict doesn't throw any
+            # errors upon syn.store, nor do they persist after Dataset refreshes
+            else:
+                raise LookupError("Extra properties found for DatasetItem: %s" %
+                                  str(dataset_item.keys() - required_keys))
+        else:
+            raise ValueError("Not a DatasetItem? %s" % str(dataset_item))
+
+    def add_items(self, dataset_items: List[Dict[str, str]]):
+        """
+        :param dataset_items: a list of dataset items
+        """
+        for dataset_item in dataset_items:
+            self.add_item(dataset_item)
+
+    def remove_item(self, item_id: str):
+        """
+        :param item_id: a single dataset item Synapse ID
+        """
+        item_id = id_of(item_id)
+        if item_id.startswith("syn"):
+            for i, curr_item in enumerate(self.properties.dataset_items):
+                if curr_item.get('entityId') == item_id:
+                    del self.properties.dataset_items[i]
+                    break
+        else:
+            raise ValueError("Not a Synapse ID: %s" % str(item_id))
+
+    def has_item(self, item_id):
+        """
+        :param item_id: a single dataset item Synapse ID
+        """
+        return any(item['entityId'] == item_id for item in self.properties.dataset_items)
```

Comment:
Can a dataset contain an entity of multiple versions? That is, can I have:

(Syn123, version1),(syn123,version2) 

In the dataset?

Reply:
No, it can't, which the API handles:

```python
synapseclient.core.exceptions.SynapseHTTPError: 400 Client Error: 
Each dataset item must have a unique entity ID.  Duplicate: syn11111111
```

Reply:
Great. Thanks!